### PR TITLE
Rework exception handling

### PR DESCRIPTION
**Pull request by PawelPawelec-RDX**

### CODE REVIEW

File: src/RadixDlt.NetworkGateway.GatewayApi/Exceptions/KnownGatewayErrorException.cs [hunk 1]
```diff
@@ -72,13 +72,13 @@ public abstract class KnownGatewayErrorException : Exception
 {
     public int StatusCode { get; }
 
-    public GatewayModel.GatewayError GatewayError { get; }
+    public GatewayModel.GatewayError? GatewayError { get; }
```
Comment (PawelPawelec-RDX) [inline src/RadixDlt.NetworkGateway.GatewayApi/Exceptions/KnownGatewayErrorException.cs:75]: Do we want to add GatewayErrors for:

- ClientClosedConnection
- BadRequestException

File: src/RadixDlt.NetworkGateway.GatewayApi/Exceptions/BadRequestException.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+﻿namespace RadixDlt.NetworkGateway.GatewayApi.Exceptions;
+
+public sealed class BadRequestException : KnownGatewayErrorException
+{
+    public BadRequestException(string userFacingMessage, string internalMessage)
+        : base(400, null, userFacingMessage, internalMessage)
```
Comment (krzlabrdx) [inline src/RadixDlt.NetworkGateway.GatewayApi/Exceptions/BadRequestException.cs:6]: 1. `(int)HttpStatusCodes.BadRequest`
2. `GenericBadRequestException`?

File: src/RadixDlt.NetworkGateway.GatewayApi/Exceptions/ClientClosedConnectionException.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+﻿namespace RadixDlt.NetworkGateway.GatewayApi.Exceptions;
+
+public sealed class ClientClosedConnectionException : KnownGatewayErrorException
+{
+    public ClientClosedConnectionException(string userFacingMessage, string internalMessage)
```
Comment (krzlabrdx) [inline src/RadixDlt.NetworkGateway.GatewayApi/Exceptions/ClientClosedConnectionException.cs:5]: `ClientConnectionClosedException` - avoid verbs in type names.

File: src/RadixDlt.NetworkGateway.GatewayApi/Exceptions/ClientClosedConnectionException.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+        : base(499, null, userFacingMessage, internalMessage)
```
Comment (krzlabrdx) [inline src/RadixDlt.NetworkGateway.GatewayApi/Exceptions/ClientClosedConnectionException.cs:6]: BTW define `private const int ClientConnectionClosedHttpStatusCode = 499` to "document" this value. Actually we might want to add some short comment / docs on why we've decided to go with non-standard 499.

File: src/RadixDlt.NetworkGateway.GatewayApi/Exceptions/InternalServerException.cs [hunk 5]
```diff
@@ -84,6 +84,17 @@ public static InternalServerException OfShareableException(Exception exception,
         );
     }
 
+    public static InternalServerException OfRequestTimeoutException(Exception exception, string traceId)
+    {
+        var message = $"Request timed out. If reporting this issue, please include TraceId={traceId}";
```
Comment (krzlabrdx) [inline src/RadixDlt.NetworkGateway.GatewayApi/Exceptions/InternalServerException.cs:89]: IIRC this "please copy trace id when reporting" message is already part of the overall response? Bit pointless to define such developer-oriented message over here.

I'd also go a little bit more into the reasoning why request has timed out, i.e. make it clear that this is our internal "max request processing duration" timeout or something.

File: src/RadixDlt.NetworkGateway.GatewayApi/Exceptions/KnownGatewayErrorException.cs (hunk 1 above)
Comment (krzlabrdx) [reply src/RadixDlt.NetworkGateway.GatewayApi/Exceptions/KnownGatewayErrorException.cs:75]: I don't think we should document such low-level / technical exceptional situations like closed connection. There's a non-negligible chance such closed connection might be short-circuited by software higher up in the overall infrastructure pipeline.

**Review by krzlabrdx — CHANGES_REQUESTED** [review]

File: apps/GatewayApi/ExceptionHandlingMiddleware/ExceptionHandlingMiddleware.cs [hunk 6]
```diff
@@ -62,60 +62,105 @@
...
+    private readonly LogLevel _knownGatewayErrorLogLevel;
 
-    public ExceptionHandler(ILogger<ExceptionHandler> logger, IHostEnvironment env, IEnumerable<IExceptionObserver> observers)
+    public ExceptionHandlingMiddleware(RequestDelegate next, IHostEnvironment env, ILogger<ExceptionHandlingMiddleware> logger,
+        IEnumerable<IExceptionObserver> observers)
     {
+        _next = next;
         _logger = logger;
         _observers = observers;
         _knownGatewayErrorLogLevel = env.IsDevelopment() ? LogLevel.Information : LogLevel.Debug;
```
Comment (krzlabrdx) [inline apps/GatewayApi/ExceptionHandlingMiddleware/ExceptionHandlingMiddleware.cs:97]: Do we even need that? I'd vote to ditch it.

**Review by krzlabrdx — CHANGES_REQUESTED** [review]

File: src/RadixDlt.NetworkGateway.GatewayApi/Exceptions/InternalServerException.cs (hunk 5 above)
Comment (PawelPawelec-RDX) [reply src/RadixDlt.NetworkGateway.GatewayApi/Exceptions/InternalServerException.cs:89]: well, that's not part of log message. That's user facing message that's returned from API response.

File: apps/GatewayApi/ExceptionHandlingMiddleware/ExceptionHandlingMiddleware.cs [hunk 7]
```diff
@@ -62,60 +62,105 @@
...
+                    return InternalServerException.OfRequestTimeoutException(exception, traceId);
+                }
+
+                if (requestAbortedFeature?.CancellationToken.IsCancellationRequested == true)
+                {
+                    _logger.LogWarning(exception, "Request aborted by user. [RequestTrace={TraceId}]", traceId);
+                    return new ClientClosedConnectionException("Client closed connection", "Client closed connection");
+                }
+
+                _logger.LogWarning(
```
[the remark this answers is not in the document]
Comment (PawelPawelec-RDX) [reply apps/GatewayApi/ExceptionHandlingMiddleware/ExceptionHandlingMiddleware.cs:157]: we log internal server errors in different case in that switch, there is no other place where internal server errors are logged.

File: src/RadixDlt.NetworkGateway.GatewayApi/Exceptions/KnownGatewayErrorException.cs (hunk 1 above)
Comment (PawelPawelec-RDX) [reply src/RadixDlt.NetworkGateway.GatewayApi/Exceptions/KnownGatewayErrorException.cs:75]: ok so lets keep that as nullable.

**Review by krzlabrdx — APPROVED** [review]

**Pull request merged**